[SPARK-20654][core] Add config to limit disk usage of the history server.

vanzin opened the pull request:
This change adds a new configuration option and support code that limits
how much disk space the SHS will use. The default value is pretty generous
so that applications will, hopefully, only rarely need to be replayed
because of their disk stored being evicted.

This works by keeping track of how much data each application is using.
Also, because it's not possible to know, before replaying, how much space
will be needed, it's possible that usage will exceed the configured limit
temporarily. The code uses the concept of a "lease" to try to limit how
much the SHS will exceed the limit in those cases.

Active UIs are also tracked, so they're never deleted. This works in
tandem with the existing option of how many active UIs are loaded; because
unused UIs will be unloaded, their disk stores will also become candidates
for deletion. If the data is not deleted, though, re-loading the UI is
pretty quick.

Commit:
[SPARK-20654][core] Add config to limit disk usage of the history server.

This change adds a new configuration option and support code that limits
how much disk space the SHS will use. The default value is pretty generous
so that applications will, hopefully, only rarely need to be replayed
because of their disk stored being evicted.

This works by keeping track of how much data each application is using.
Also, because it's not possible to know, before replaying, how much space
will be needed, it's possible that usage will exceed the configured limit
temporarily. The code uses the concept of a "lease" to try to limit how
much the SHS will exceed the limit in those cases.

Active UIs are also tracked, so they're never deleted. This works in
tandem with the existing option of how many active UIs are loaded; because
unused UIs will be unloaded, their disk stores will also become candidates
for deletion. If the data is not deleted, though, re-loading the UI is
pretty quick.

SparkQA commented:
**[Test build #85072 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85072/testReport)** for PR 20011 at commit [`8b43f4d`](https://github.com/apache/spark/commit/8b43f4d2c06ab0a9916113fc216711573db367e7).

vanzin commented:
Some previous comments for this PR can be found at: https://github.com/vanzin/spark/pull/39

SparkQA commented:
**[Test build #85072 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85072/testReport)** for PR 20011 at commit [`8b43f4d`](https://github.com/apache/spark/commit/8b43f4d2c06ab0a9916113fc216711573db367e7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85072/
Test PASSed.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/DiskStoreManager.scala:
```diff
@@ -0,0 +1,285 @@
+   * event log. By default it's 30% of the event log size.
+   */
+  private def approximateSize(eventLogSize: Long): Long = {
+    math.ceil(eventLogSizeRatio * eventLogSize).toLong
```
I'll probably need to look at a better heuristic here. This is probably not a good approximation if the logs are compressed, and #20013 also affects this.

Commit:
Change approximate size heuristic to account for compression.

Also remove the config to tweak the heuristic since it doesn't make
a lot of sense anymore.

SparkQA commented:
**[Test build #85129 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85129/testReport)** for PR 20011 at commit [`31c2755`](https://github.com/apache/spark/commit/31c2755f5f4f04966c9b0061617563024b6a855f).

SparkQA commented:
**[Test build #85129 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85129/testReport)** for PR 20011 at commit [`31c2755`](https://github.com/apache/spark/commit/31c2755f5f4f04966c9b0061617563024b6a855f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85129/
Test PASSed.

Commit:
Merge branch 'master' into SPARK-20654

SparkQA commented:
**[Test build #85284 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85284/testReport)** for PR 20011 at commit [`2d72dd1`](https://github.com/apache/spark/commit/2d72dd1d08c5fcb174dea14dd0586e1f3f6d4c05).

SparkQA commented:
**[Test build #85284 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85284/testReport)** for PR 20011 at commit [`2d72dd1`](https://github.com/apache/spark/commit/2d72dd1d08c5fcb174dea14dd0586e1f3f6d4c05).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85284/
Test PASSed.

squito commented on core/src/test/scala/org/apache/spark/deploy/history/DiskStoreManagerSuite.scala:
```diff
@@ -0,0 +1,161 @@
+    val lease5 = manager.lease(1)
+    doReturn(3L).when(manager).sizeOf(meq(lease5.path))
+    lease5.commit("app2", None)
+    assert(dst2.exists())
```
would help here to have a comment explaining the case this really represents -- replaying an app because the event log has been updated, so replacing the old kvstore.  Also just to make that clear you could add

```
val dst5 = lease5.commit("app2", None)
assert(dst5 === dst2)
```
(at first I thought "app2" was a copy/paste mistake here)

squito commented on core/src/test/scala/org/apache/spark/deploy/history/DiskStoreManagerSuite.scala:
```diff
@@ -0,0 +1,161 @@
+    // Try a big lease that should cause the committed app to be evicted.
+    val lease6 = manager.lease(6)
+    assert(!dst2.exists())
+    assert(!hasFreeSpace(manager, 1))
```
add a case with getting another lease, even though you're already beyond the size limit (don't think that is covered?)

squito commented on core/src/main/scala/org/apache/spark/deploy/history/DiskStoreManager.scala:
```diff
@@ -0,0 +1,296 @@
+  /** Visible for testing. Return the size of a directory. */
+  private[history] def sizeOf(path: File): Long = FileUtils.sizeOf(path)
+
+  private[history] class Lease(val path: File, private val leased: Long) {
```
can you rename path to tmpPath

squito commented on core/src/main/scala/org/apache/spark/deploy/history/DiskStoreManager.scala:
```diff
@@ -0,0 +1,296 @@
+
+      active.synchronized {
+        require(!active.contains(appId -> attemptId),
+          s"Cannot commit lease for active application $appId / $attemptId")
```
what is the scenario you're worried about here?  you could have two threads both trying to call commit, and they might both get past this check, and then clobber each other later on, so this is more limited protection.  The idea is just to make sure that you never do a commit while the UI has a store open, as a check of the whole UI cache eviction process?

vanzin replied:
This is more of a sanity check that the situation you describe should not happen. The SHS code should be ensuring that there's only a single thread parsing logs for an application, and that while that happens, the app's UI is not loaded. This just asserts that's the case.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/DiskStoreManager.scala:
```diff
@@ -0,0 +1,296 @@
+
+    // Cap the value at 10% of the max size; this assumes that element cleanup will put a cap on
+    // how large the disk store can get, which may not always be the case.
+    math.min(expectedSize, maxUsage / 10)
```
Well, all of these heuristics can have a lot of error. Increasing this increases the odds that an existing store will be deleted if another log starts to be parsed in parallel.

Don't really feel strongly either way, so away it goes.

vanzin commented on core/src/test/scala/org/apache/spark/deploy/history/DiskStoreManagerSuite.scala:
```diff
@@ -0,0 +1,161 @@
+  private def mockManager(): DiskStoreManager = {
+    val conf = new SparkConf().set(MAX_LOCAL_DISK_USAGE, MAX_USAGE)
+    val manager = spy(new DiskStoreManager(conf, testDir, store, new ManualClock()))
+    doReturn(0L).when(manager).sizeOf(any(classOf[File]))
```
Just added in case the method was called with an unexpected file. Will remove.

Commit:
Feedback.

Commit:
Rename class.

Commit:
Track committed usage separately to make log message more accurate.

Also fix a couple of bugs found in the process.

SparkQA commented:
**[Test build #85320 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85320/testReport)** for PR 20011 at commit [`be3f130`](https://github.com/apache/spark/commit/be3f1307f0edffdc7c9457ec960781cd28b07bf8).

squito commented on core/src/main/scala/org/apache/spark/deploy/history/HistoryServerDiskManager.scala:
```diff
@@ -0,0 +1,310 @@
+        logInfo(s"Deleting store for ${info.appId}/${info.attemptId}.")
+        deleteStore(new File(info.path))
+        updateUsage(-info.size, committed = true)
+      }
```
I like the change to track committed space.  I'm wondering if we should log more here, even at Info level, summarizing the available space (I can see users complaining that they configured it to 10GB but its taking 12GB and wanting an explanation -- would be nice if the logs already had all the relevant info).  Maybe something like

```
val totalNeeded = size
var activeSize = 0
...
if (!isActive) {
  evicted += info
  needed -= info.size
} else {
  activeSize += info.size
}
...
logInfo(s"Deleted ${evicted.size} apps to free ${Utils.bytesToString(needed)})
if (needed > 0) {
  val current = currentUsage.get()
  val leased = Utils.bytesToString(current - committedUsage.get())
  logInfo(s"Did not free requested ${Utils.bytesToString(totalNeeded)}.  Current usage is ${Utils.bytesToString(current).  $leased (estimated) used by apps actively updating their kvstores; ${Utils.bytesToString(active)} used by active applications.")
}

vanzin replied:
Yeah, I'll add some more logs.

Commit:
Add more logs.

SparkQA commented:
**[Test build #85326 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85326/testReport)** for PR 20011 at commit [`931b2d2`](https://github.com/apache/spark/commit/931b2d262aa02880631ca4c693a84fa4c4d12318).

SparkQA commented:
**[Test build #85320 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85320/testReport)** for PR 20011 at commit [`be3f130`](https://github.com/apache/spark/commit/be3f1307f0edffdc7c9457ec960781cd28b07bf8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85320/
Test PASSed.

SparkQA commented:
**[Test build #85326 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85326/testReport)** for PR 20011 at commit [`931b2d2`](https://github.com/apache/spark/commit/931b2d262aa02880631ca4c693a84fa4c4d12318).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85326/
Test PASSed.

squito reviewed (approved):
lgtm

squito commented:
merged to master
